Name: Zenas Huang <br>
USC ID: 8806-6035-51 <br>
Class: ISE 529 - Predictive Analytics <br>
Section: Wednesday, 5 - 7:50PM <br>
Term: Fall 2019 Semester <br>
Instructor: Dr. Cesar Acosta-Mejia

# Homework 2 #

A real estate appraiser is interested in predicting residential home prices as a function of various features. Therefore a regression model is to be constructed from a sample of 522 houses. Download the homes.csv data set from blackboard. Substitute the variable year with a new variable age (assume that current year is 2001).



Consider the numeric predictors x1: lot size (square feet), x2: area (square feet), x3: number of bedrooms, x4: number of bathrooms, x5: age, and, x6: garage size (number of cars).



In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df_homes = pd.read_csv('homes.csv')

In [3]:
df_homes.head()

price  area  beds  baths  garage  year  style  lotsize   ac pool quality  \
0  360000  3032     4      4       2  1972      1    22221  YES   NO  MEDIUM   
1  340000  2058     4      2       2  1976      1    22912  YES   NO  MEDIUM   
2  250000  1780     4      3       2  1980      1    21345  YES   NO  MEDIUM   
3  205500  1638     4      2       2  1963      1    17342  YES   NO  MEDIUM   
4  275500  2196     4      3       2  1968      7    21786  YES   NO  MEDIUM   

  highway  
0      NO  
1      NO  
2      NO  
3      NO  
4      NO

In [4]:
#add a variable "Age" to the dataframe on assumption that 
#the current year is 2001

In [5]:
df_homes['age'] = 2001 - df_homes['year']

In [6]:
df_homes.head()

price  area  beds  baths  garage  year  style  lotsize   ac pool quality  \
0  360000  3032     4      4       2  1972      1    22221  YES   NO  MEDIUM   
1  340000  2058     4      2       2  1976      1    22912  YES   NO  MEDIUM   
2  250000  1780     4      3       2  1980      1    21345  YES   NO  MEDIUM   
3  205500  1638     4      2       2  1963      1    17342  YES   NO  MEDIUM   
4  275500  2196     4      3       2  1968      7    21786  YES   NO  MEDIUM   

  highway  age  
0      NO   29  
1      NO   25  
2      NO   21  
3      NO   38  
4      NO   33

In [7]:
#target variable is home price
#predictor is number of bedrooms
df_homes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 522 entries, 0 to 521
Data columns (total 13 columns):
price      522 non-null int64
area       522 non-null int64
beds       522 non-null int64
baths      522 non-null int64
garage     522 non-null int64
year       522 non-null int64
style      522 non-null int64
lotsize    522 non-null int64
ac         522 non-null object
pool       522 non-null object
quality    522 non-null object
highway    522 non-null object
age        522 non-null int64
dtypes: int64(9), object(4)
memory usage: 53.1+ KB


In [8]:
#ac, pool, highway are all NO, YES variables
df_homes.quality.unique()

array(['MEDIUM', 'LOW', 'HIGH'], dtype=object)

## Question 1
Fit a model with one predictor only (the number of bedrooms). What is the estimated increase in the average price of a house for each additional bedroom?

In [9]:
X3 = df_homes.beds
price = df_homes["price"]

In [10]:
from sklearn.linear_model import LinearRegression

In [11]:
#fit the model

In [12]:
X3.shape

(522,)

In [13]:
X3 = X3.values.reshape(-1,1)
X3.shape

(522, 1)

In [14]:
m2 = LinearRegression().fit(X3, price)
m2.fit(X3,price)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [15]:
m2.intercept_

82808.79961261936

In [16]:
m2.coef_

array([56200.08366568])

***Q1 Answer:***

The average price of a house increases by \\$56200.08 for each additional bedroom

## Question 2
Fit the full model (with variables x1,...,x6)

We note that each of the variables x1, ..., x6 are all already numeric valued.

x1 = lot size, x2 = area, x3 = number of bedrooms, x4 = number of bathrooms, x5 = age, x6 = garage size (number of cars)

### 2 a) 
Report the estimated regression coefficients (include a 99% confidence interval for each one). 

In [17]:
df_homes.head()

price  area  beds  baths  garage  year  style  lotsize   ac pool quality  \
0  360000  3032     4      4       2  1972      1    22221  YES   NO  MEDIUM   
1  340000  2058     4      2       2  1976      1    22912  YES   NO  MEDIUM   
2  250000  1780     4      3       2  1980      1    21345  YES   NO  MEDIUM   
3  205500  1638     4      2       2  1963      1    17342  YES   NO  MEDIUM   
4  275500  2196     4      3       2  1968      7    21786  YES   NO  MEDIUM   

  highway  age  
0      NO   29  
1      NO   25  
2      NO   21  
3      NO   38  
4      NO   33

In [18]:
#select variables

In [19]:
list1 = ['price','lotsize', 'area', 'beds', 'baths', 'age', 'garage']
list1

['price', 'lotsize', 'area', 'beds', 'baths', 'age', 'garage']

In [20]:
df1 = df_homes[list1]
df1[:5]

price  lotsize  area  beds  baths  age  garage
0  360000    22221  3032     4      4   29       2
1  340000    22912  2058     4      2   25       2
2  250000    21345  1780     4      3   21       2
3  205500    17342  1638     4      2   38       2
4  275500    21786  2196     4      3   33       2

In [21]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 522 entries, 0 to 521
Data columns (total 7 columns):
price      522 non-null int64
lotsize    522 non-null int64
area       522 non-null int64
beds       522 non-null int64
baths      522 non-null int64
age        522 non-null int64
garage     522 non-null int64
dtypes: int64(7)
memory usage: 28.6 KB


In [22]:
df1.corr()

price   lotsize      area      beds     baths       age    garage
price    1.000000  0.224169  0.819470  0.413324  0.683685 -0.555516  0.577786
lotsize  0.224169  1.000000  0.157525  0.126538  0.147007  0.100452  0.152219
area     0.819470  0.157525  1.000000  0.557838  0.755273 -0.441197  0.533766
beds     0.413324  0.126538  0.557838  1.000000  0.583447 -0.268692  0.316814
baths    0.683685  0.147007  0.755273  0.583447  1.000000 -0.512841  0.489898
age     -0.555516  0.100452 -0.441197 -0.268692 -0.512841  1.000000 -0.461760
garage   0.577786  0.152219  0.533766  0.316814  0.489898 -0.461760  1.000000

In [23]:
#select predictor columns and response
X0 = df1.iloc[:,1:7]
X0[:5]

lotsize  area  beds  baths  age  garage
0    22221  3032     4      4   29       2
1    22912  2058     4      2   25       2
2    21345  1780     4      3   21       2
3    17342  1638     4      2   38       2
4    21786  2196     4      3   33       2

In [24]:
#select response variable
y0 = df_homes.iloc[:,0]
y0[:5]

0    360000
1    340000
2    250000
3    205500
4    275500
Name: price, dtype: int64

In [25]:
# Fit the Multiple Linear Regression Model

In [26]:
import statsmodels.api as sm

In [27]:
X0.shape

(522, 6)

In [28]:
#convert X0 to numpy array 
X = X0.to_numpy()
X[:5]

array([[22221,  3032,     4,     4,    29,     2],
       [22912,  2058,     4,     2,    25,     2],
       [21345,  1780,     4,     3,    21,     2],
       [17342,  1638,     4,     2,    38,     2],
       [21786,  2196,     4,     3,    33,     2]])

In [29]:
X1 = sm.add_constant(X)
X1[:5]

array([[1.0000e+00, 2.2221e+04, 3.0320e+03, 4.0000e+00, 4.0000e+00,
        2.9000e+01, 2.0000e+00],
       [1.0000e+00, 2.2912e+04, 2.0580e+03, 4.0000e+00, 2.0000e+00,
        2.5000e+01, 2.0000e+00],
       [1.0000e+00, 2.1345e+04, 1.7800e+03, 4.0000e+00, 3.0000e+00,
        2.1000e+01, 2.0000e+00],
       [1.0000e+00, 1.7342e+04, 1.6380e+03, 4.0000e+00, 2.0000e+00,
        3.8000e+01, 2.0000e+00],
       [1.0000e+00, 2.1786e+04, 2.1960e+03, 4.0000e+00, 3.0000e+00,
        3.3000e+01, 2.0000e+00]])

In [30]:
m2 = sm.OLS(y0,X1).fit()

In [31]:
m2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.751
Model:                            OLS   Adj. R-squared:                  0.748
Method:                 Least Squares   F-statistic:                     259.4
Date:                Wed, 25 Sep 2019   Prob (F-statistic):          4.21e-152
Time:                        16:19:19   Log-Likelihood:                -6554.5
No. Observations:                 522   AIC:                         1.312e+04
Df Residuals:                     515   BIC:                         1.315e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -6707.5382   1.88e+04     -0.356      0.722   -4.37e+04    3.03e+04
x1             1.5550      0.272      5.717      0.000       1.021       2.089
x2           125.7386      6.963     18.057      0.000     112.059     139.419
x3         -1.304e+04   3780.397     -3.450      0.001   -2.05e+04   -5614.498
x4          7987.5518   4818.957      1.658      0.098   -1479.680    1.75e+04
x5         -1779.6108    215.487     -8.259      0.000   -2202.953   -1356.268
x6          2.253e+04   5814.344      3.875      0.000    1.11e+04     3.4e+04
==============================================================================
Omnibus:                      124.114   Durbin-Watson:                   1.327
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              384.582
Skew:                           1.110   Prob(JB):                     3.08e-84
Kurtosis:                       6.571   Cond. No.                     1.71e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.71e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [32]:
m2.params

const    -6707.538167
x1           1.554990
x2         125.738603
x3      -13041.394214
x4        7987.551827
x5       -1779.610838
x6       22530.379289
dtype: float64

##### The Full Fitted Equation
yhat = -6707.5382 + 1.55499 lotsize + 125.7386 area - 13041.3942 beds + 7987.5518 baths -1779.6108 age + 22530.3793 garage


***Q2a Answer:***

99% Confidence Intervals for each predictors. We are 99% confident that the value of each coefficient lies within each of the given intervals below.

In [33]:
m2.conf_int(alpha=0.01)

0             1
const -55431.280841  42016.204508
x1         0.851721      2.258260
x2       107.735683    143.741523
x3    -22815.268707  -3267.519722
x4     -4471.425313  20446.528968
x5     -2336.734049  -1222.487627
x6      7497.919937  37562.838640

### 2 b)
Find a 95% confidence interval for the average price of a house with garage for two cars, area of 2650 square feet, built in 1990, 24500 square feet size, three bedroom, and three bathrooms.

 Given the assumption that the current year is 2001, for a house built in 1990, the age of the house will be 11 years old. We are asked to predict the expected price of a house with a 2-car garage, Area of 2650 square feet, and a lotsize of 24500 squarefeet, and 3 bedrooms and 3 bathrooms and then to report on a 95% confidence interval for this expected price.

In [34]:
list1

['price', 'lotsize', 'area', 'beds', 'baths', 'age', 'garage']

In [35]:
#remove response name

In [36]:
list1.remove('price')
list1

['lotsize', 'area', 'beds', 'baths', 'age', 'garage']

In [37]:
list1.insert(0,'constant')
list1

['constant', 'lotsize', 'area', 'beds', 'baths', 'age', 'garage']

In [38]:
list_values = [1,24500, 2650, 3, 3, 11, 2]

In [39]:
x = pd.DataFrame([list_values], columns = list1)
x

constant  lotsize  area  beds  baths  age  garage
0         1    24500  2650     3      3   11       2

In [40]:
d2 = m2.get_prediction(x)

In [41]:
d2.summary_frame(alpha = .05)

mean      mean_se  mean_ci_lower  mean_ci_upper  obs_ci_lower  \
0  374920.532647  6511.365875  362128.426983   387712.63831  238431.17994   

    obs_ci_upper  
0  511409.885353

***Q2b Answer:*** The 95% confidence interval for this house's expected price is [ \\$362128, \\$387712 ].

### 2 c)

Predict the price of a house when all predictors are equal to their median values.

In [42]:
X0.head()

lotsize  area  beds  baths  age  garage
0    22221  3032     4      4   29       2
1    22912  2058     4      2   25       2
2    21345  1780     4      3   21       2
3    17342  1638     4      2   38       2
4    21786  2196     4      3   33       2

In [43]:
newval = X0.iloc[[1]]
newval

lotsize  area  beds  baths  age  garage
1    22912  2058     4      2   25       2

In [44]:
newval.iat[0,0] = np.median(X0.lotsize)
newval.iat[0,1] = np.median(X0.area)
newval.iat[0,2] = np.median(X0.beds)
newval.iat[0,3] = np.median(X0.baths)
newval.iat[0,4] = np.median(X0.age)
newval.iat[0,5] = np.median(X0.garage)
newval

lotsize  area  beds  baths  age  garage
1    22200  2061     3      3   35       2

In [45]:
list_values_median = [1,22200, 2061, 3, 3, 35, 2]

In [46]:
m = pd.DataFrame([list_values_median], columns = list1)
m

constant  lotsize  area  beds  baths  age  garage
0         1    22200  2061     3      3   35       2

In [47]:
d3 = m2.get_prediction(m)

In [48]:
d3.predicted_mean

array([254573.35791857])

In [49]:
d3.summary_frame(alpha = .05)

mean      mean_se  mean_ci_lower  mean_ci_upper   obs_ci_lower  \
0  254573.357919  4595.921615   245544.29765  263602.418187  118385.144904   

    obs_ci_upper  
0  390761.570933

***Q2c Answer:*** The predicted price for a house with all median values will be \\$254,573.36

## Question 3

What are the best and worst predictors in terms of correlation with the house price?

In [50]:
corr = df_homes.corr()
corr

price      area      beds     baths    garage      year     style  \
price    1.000000  0.819470  0.413324  0.683685  0.577786  0.555516  0.357493   
area     0.819470  1.000000  0.557838  0.755273  0.533766  0.441197  0.616842   
beds     0.413324  0.557838  1.000000  0.583447  0.316814  0.268692  0.380370   
baths    0.683685  0.755273  0.583447  1.000000  0.489898  0.512841  0.492983   
garage   0.577786  0.533766  0.316814  0.489898  1.000000  0.461760  0.234848   
year     0.555516  0.441197  0.268692  0.512841  0.461760  1.000000  0.234059   
style    0.357493  0.616842  0.380370  0.492983  0.234848  0.234059  1.000000   
lotsize  0.224169  0.157525  0.126538  0.147007  0.152219 -0.100452 -0.013115   
age     -0.555516 -0.441197 -0.268692 -0.512841 -0.461760 -1.000000 -0.234059   

          lotsize       age  
price    0.224169 -0.555516  
area     0.157525 -0.441197  
beds     0.126538 -0.268692  
baths    0.147007 -0.512841  
garage   0.152219 -0.461760  
year    -0.100452 -1.000000  
style   -0.013115 -0.234059  
lotsize  1.000000  0.100452  
age      0.100452  1.000000

In [51]:
abs_corr = df_homes.corr().abs()
abs_corr.iloc[0]

price      1.000000
area       0.819470
beds       0.413324
baths      0.683685
garage     0.577786
year       0.555516
style      0.357493
lotsize    0.224169
age        0.555516
Name: price, dtype: float64

***Q3 Answer:*** 

Best is Area.

Worst is Lotsize.

The best predictors in terms of correlation is the one with the highest absolute value correlation with the response variable price. In this case, the best is: Area (0.81).

The worst predictors in terms of correlation are those with the smallest absolute value correlation with the response (price). So, the worst is: lotsize (0.22)

## Question 4

What are the best and worst predictors in terms of R-squared?

*Addendum Note 1 (9/23/19):* It was clarified by the TA that this may be a comparison of R^2 values among all single variable regressions. 


*Addendum Note 2 (9/24/19):* Further re-clarification by Professor Acosta: this question is *not* asking for a comparison among R^2 values for univariate regressions on each predictor but for a comparison among the combinatorial ranking of models as done in lecture. The predictor with the highest frequency in all the model subsets is the best predictor and the predictor appearing with the lowest frequency in all the model subsets is the worst predictor in terms of R-squared. A comparison of R^2's among the univariate regressions will be superfluous since R^2 is the square of the correlation coefficient values, so it would just yield the exact same result as the absolute value correlation based ranking of worst and best predictors.

In [52]:
from sklearn.metrics import mean_squared_error
from scipy.special import comb
import itertools

In [53]:
def get_sse(X,Y):
    model_k = LinearRegression().fit(X,Y)
    SSE = mean_squared_error(Y,model_k.predict(X))*len(Y)
    R_squared = model_k.score(X,Y)
    return SSE, R_squared

In [54]:
SSE, R_squared = get_sse(X0,y0)

In [55]:
SSE

2463957738779.1533

In [56]:
R_squared

0.7513894013091613

In [57]:
#Create four empty lists

In [58]:
SSE_list, R2_list, feature_list, num_features = [],[],[],[]

In [59]:
col_names = X0.columns
col_names

Index(['lotsize', 'area', 'beds', 'baths', 'age', 'garage'], dtype='object')

In [60]:
p = len(X0.columns)
p

6

In [61]:
list(itertools.combinations(col_names,2))

[('lotsize', 'area'),
 ('lotsize', 'beds'),
 ('lotsize', 'baths'),
 ('lotsize', 'age'),
 ('lotsize', 'garage'),
 ('area', 'beds'),
 ('area', 'baths'),
 ('area', 'age'),
 ('area', 'garage'),
 ('beds', 'baths'),
 ('beds', 'age'),
 ('beds', 'garage'),
 ('baths', 'age'),
 ('baths', 'garage'),
 ('age', 'garage')]

In [62]:
list(range(1,p+1))

[1, 2, 3, 4, 5, 6]

In [63]:
for k in range(1,p+1):
    for subset in itertools.combinations(X0.columns,k):
        temp = get_sse(X0[list(subset)],y0)
        
        SSE_list.append(temp[0])
        R2_list.append(temp[1])
        feature_list.append(subset)
        num_features.append(len(subset))

In [64]:
#Create dataframe with 4 cols
df = pd.DataFrame(num_features,columns=['n_features'])

In [65]:
SSE_list[:5]

[9412873469063.32,
 3255425605236.3877,
 8217765186285.018,
 5278296874758.204,
 6852419094443.968]

In [66]:
SSE_list2 = np.round(SSE_list,0)

In [67]:
df['features'] = feature_list
df['R-squared'] = R2_list
df['SSE'] = SSE_list2

In [68]:
best_r2 = df.groupby(['n_features'])['R-squared'].max()

In [69]:
best_r2

n_features
1    0.671531
2    0.718249
3    0.737944
4    0.745438
5    0.750063
6    0.751389
Name: R-squared, dtype: float64

In [70]:
#dataframe with best single predictor
df2 = df[df['R-squared']== best_r2[1]]
df2

n_features features  R-squared           SSE
1           1  (area,)   0.671531  3.255426e+12

In [71]:
#dataframe with best two predictors
df2 = df[df['R-squared']== best_r2[2]]
df2

n_features     features  R-squared           SSE
13           2  (area, age)   0.718249  2.792410e+12

In [72]:
#dataframe with best subsets of predictors

In [73]:
df2 = df[df['R-squared']== best_r2[1]]

In [74]:
df2

n_features features  R-squared           SSE
1           1  (area,)   0.671531  3.255426e+12

In [75]:
for i in range(2,7):
    df2 = df2.append(df[df['R-squared'] == best_r2[i]])

In [76]:
#pd.set_option('display.max_colwidth',190)
df2

n_features                                   features  R-squared  \
1            1                                    (area,)   0.671531   
13           2                                (area, age)   0.718249   
23           3                       (lotsize, area, age)   0.737944   
46           4               (lotsize, area, age, garage)   0.745438   
58           5         (lotsize, area, beds, age, garage)   0.750063   
62           6  (lotsize, area, beds, baths, age, garage)   0.751389   

             SSE  
1   3.255426e+12  
13  2.792410e+12  
23  2.597213e+12  
46  2.522941e+12  
58  2.477102e+12  
62  2.463958e+12

***Q4 Answer:***

In terms of R-squared, Area is the best predictor. Baths is the worst because it appears last in the list of our combinatorial ranking. This was verified as correct during Office Hours by Professor Acosta (9/24/2019).

## Question 5

Fit best model with best subset of predictors (in terms of AIC). Report the fitted equation.

In [77]:
n = len(df_homes)

In [78]:
#Add AIC and adjusted R-squared columns
df2['AIC'] = n*np.log(df2['SSE']/n)+2*df2['n_features']
df2['adj_R-squared']=1-((1-df2['R-squared'])*(n-1)/(n-df2['n_features']-1))

In [79]:
df2

n_features                                   features  R-squared  \
1            1                                    (area,)   0.671531   
13           2                                (area, age)   0.718249   
23           3                       (lotsize, area, age)   0.737944   
46           4               (lotsize, area, age, garage)   0.745438   
58           5         (lotsize, area, beds, age, garage)   0.750063   
62           6  (lotsize, area, beds, baths, age, garage)   0.751389   

             SSE           AIC  adj_R-squared  
1   3.255426e+12  11775.019159       0.670900  
13  2.792410e+12  11696.934938       0.717163  
23  2.597213e+12  11661.107762       0.736426  
46  2.522941e+12  11647.962610       0.743468  
58  2.477102e+12  11640.391242       0.747641  
62  2.463958e+12  11639.613904       0.748493

In [80]:
#AIC Best Model
df2.AIC.min()

11639.613904336478

In [81]:
df2[df2.AIC == df2.AIC.min()]

n_features                                   features  R-squared  \
62           6  (lotsize, area, beds, baths, age, garage)   0.751389   

             SSE           AIC  adj_R-squared  
62  2.463958e+12  11639.613904       0.748493

In [82]:
min_AIC = df2['AIC'].min()
min_AIC

11639.613904336478

In [83]:
row6 = df2[df2['AIC'] == min_AIC]

In [84]:
row6['features']

62    (lotsize, area, beds, baths, age, garage)
Name: features, dtype: object

In [85]:
list2 = ['lotsize', 'area', 'beds', 'baths', 'age', 'garage']
list2

['lotsize', 'area', 'beds', 'baths', 'age', 'garage']

In [86]:
X_aic = X0[list2]

In [87]:
m3 = LinearRegression().fit(X_aic,y0)

In [88]:
m3.score(X_aic,y0)

0.7513894013091613

In [89]:
[(round(n,2)) for n in m3.coef_]

[1.55, 125.74, -13041.39, 7987.55, -1779.61, 22530.38]

In [90]:
round(m3.intercept_,2)

-6707.54

***Q5 Answer:***

The fitted best AIC model is:

yhat = -6707.54 + 1.55 lotsize + 125.74 area -13041.39 beds + 7987.55 baths - 1779.61 age + 22530.38 garage

Consider now predictors lotsize and quality only

## Question 6

Fit a linear regression model and report the fitted equations for lotsize and quality only.

In [91]:
df_homes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 522 entries, 0 to 521
Data columns (total 13 columns):
price      522 non-null int64
area       522 non-null int64
beds       522 non-null int64
baths      522 non-null int64
garage     522 non-null int64
year       522 non-null int64
style      522 non-null int64
lotsize    522 non-null int64
ac         522 non-null object
pool       522 non-null object
quality    522 non-null object
highway    522 non-null object
age        522 non-null int64
dtypes: int64(9), object(4)
memory usage: 53.1+ KB


In [92]:
#we note that quality is a categorical variable, so we must first encode it
df_homes['quality'].unique()

array(['MEDIUM', 'LOW', 'HIGH'], dtype=object)

In [93]:
df_homes1 = pd.get_dummies(df_homes, columns = ['quality'], drop_first = True)

In [94]:
df_homes1[:10]

price  area  beds  baths  garage  year  style  lotsize   ac pool highway  \
0  360000  3032     4      4       2  1972      1    22221  YES   NO      NO   
1  340000  2058     4      2       2  1976      1    22912  YES   NO      NO   
2  250000  1780     4      3       2  1980      1    21345  YES   NO      NO   
3  205500  1638     4      2       2  1963      1    17342  YES   NO      NO   
4  275500  2196     4      3       2  1968      7    21786  YES   NO      NO   
5  248000  1966     4      3       5  1972      1    18902  YES  YES      NO   
6  229900  2216     3      2       2  1972      7    18639  YES   NO      NO   
7  150000  1597     2      1       1  1955      1    22112  YES   NO      NO   
8  195000  1622     3      2       2  1975      1    14321  YES   NO      NO   
9  160000  1976     3      3       1  1918      1    32358   NO   NO      NO   

   age  quality_LOW  quality_MEDIUM  
0   29            0               1  
1   25            0               1  
2   21            0               1  
3   38            0               1  
4   33            0               1  
5   29            0               1  
6   29            0               1  
7   46            0               1  
8   26            1               0  
9   83            1               0

In [95]:
# Base case is High Quality: Quality_low = 0 && Quality_Medium = 0
#Low Quality if Quality_Low = 1
#Medium Quality if Quality_Medium = 1
#High Quality if Quality_Low = 0 && Quality_Medium = 0

In [96]:
list3 = ['lotsize', 'quality_LOW','quality_MEDIUM']
list3

['lotsize', 'quality_LOW', 'quality_MEDIUM']

In [97]:
X_2factors = df_homes1[list3]

In [98]:
m4 = LinearRegression().fit(X_2factors,y0)

In [99]:
m4.score(X_2factors,y0)

0.6791480605904724

In [100]:
[(round(n,2)) for n in m4.coef_]

[1.64, -361721.64, -266554.19]

In [101]:
round(m4.intercept_,2)

499607.82

***Q6 Answer:***

The complete fitted regression equation with all estimators for expected price as a function of lotsize and quality is:

yhat = 499607.82 + 1.64 lotsize -361721.64 quality_LOW -266554.19 quality_MEDIUM

But since the question asks for plural fitted equation**s**, we can make separate this out for each quality group by turning on each associated dummy:

**high quality:**

yhat_high = 499607.82 + 1.64 lotsize

**medium quality:**

yhat_medium = 233053.63 + 1.64 lotsize 

**low quality:**

yhat_low = 137886.18 + 1.64 lotsize

## Question 7

Predict the price of a house with lotsize 40000 square feet and medium quality.

In [102]:
newval = X_2factors.iloc[[1]]
newval

lotsize  quality_LOW  quality_MEDIUM
1    22912            0               1

In [103]:
newval.iat[0,0] = 40000
newval.iat[0,1] = 0
newval.iat[0,2] = 1
newval

lotsize  quality_LOW  quality_MEDIUM
1    40000            0               1

In [104]:
m4.predict(newval)

array([298735.73263702])

***Q7 Answer:***

Using this two factor model, the predicted price for a house with lotsize area 40000 and of medium quality is \\$298735.73

Consider now all variables in the csv file.

## Question 8
Fit a full model for houses having between two to four bedrooms. Find a 95% prediction interval for the price of a high quality house with a garage for two cars, area of 3150 square feet, built in 1996, 26250 square feet size, two bedrooms, three bathrooms, style 1, with AC, a pool, and not close to a highway.

First we begin by encoding all of the categorical variables in the CSV File. We note that although style is "numeric" valued data it is in fact an identifying state, as such it is a categorical variable identifying a house as being in any of 10 styles (there is no home with style state 8 in the dataset). Next, for a home built in 1996, then the home will be age 5 years based on the assumption that the current year is 2001.

In [105]:
df_homes.head()

price  area  beds  baths  garage  year  style  lotsize   ac pool quality  \
0  360000  3032     4      4       2  1972      1    22221  YES   NO  MEDIUM   
1  340000  2058     4      2       2  1976      1    22912  YES   NO  MEDIUM   
2  250000  1780     4      3       2  1980      1    21345  YES   NO  MEDIUM   
3  205500  1638     4      2       2  1963      1    17342  YES   NO  MEDIUM   
4  275500  2196     4      3       2  1968      7    21786  YES   NO  MEDIUM   

  highway  age  
0      NO   29  
1      NO   25  
2      NO   21  
3      NO   38  
4      NO   33

In [106]:
df_homes.beds.unique()

array([4, 3, 2, 7, 5, 1, 6, 0])

In [107]:
df_homes = df_homes[(df_homes['beds'] >= 2) & (df_homes['beds'] <= 4)]

In [131]:
#filtered data to only houses between 2-4 bedrooms
df_homes.shape

(445, 20)

In [109]:
df_homes = pd.get_dummies(df_homes, columns = ['style', 'ac', 'pool','quality','highway'], drop_first = True)

In [110]:
df_homes.head()

price  area  beds  baths  garage  year  lotsize  age  style_2  style_3  \
0  360000  3032     4      4       2  1972    22221   29        0        0   
1  340000  2058     4      2       2  1976    22912   25        0        0   
2  250000  1780     4      3       2  1980    21345   21        0        0   
3  205500  1638     4      2       2  1963    17342   38        0        0   
4  275500  2196     4      3       2  1968    21786   33        0        0   

   style_4  style_5  style_6  style_7  style_9  ac_YES  pool_YES  quality_LOW  \
0        0        0        0        0        0       1         0            0   
1        0        0        0        0        0       1         0            0   
2        0        0        0        0        0       1         0            0   
3        0        0        0        0        0       1         0            0   
4        0        0        0        1        0       1         0            0   

   quality_MEDIUM  highway_YES  
0               1            0  
1               1            0  
2               1            0  
3               1            0  
4               1            0

In [113]:
#select variables
list4 = ['price', 'area', 'beds', 'baths', 'garage', 'lotsize', 'age', 'style_2', 'style_3', 'style_4', 'style_5', 'style_6', 'style_7', 'style_9', 'ac_YES', 'pool_YES', 'quality_LOW', 'quality_MEDIUM', 'highway_YES']

In [116]:
list4.remove('price')
list4

['area',
 'beds',
 'baths',
 'garage',
 'lotsize',
 'age',
 'style_2',
 'style_3',
 'style_4',
 'style_5',
 'style_6',
 'style_7',
 'style_9',
 'ac_YES',
 'pool_YES',
 'quality_LOW',
 'quality_MEDIUM',
 'highway_YES']

In [117]:
df_homes2 = df_homes[list4]
df_homes2

area  beds  baths  garage  lotsize  age  style_2  style_3  style_4  \
0    3032     4      4       2    22221   29        0        0        0   
1    2058     4      2       2    22912   25        0        0        0   
2    1780     4      3       2    21345   21        0        0        0   
3    1638     4      2       2    17342   38        0        0        0   
4    2196     4      3       2    21786   33        0        0        0   
5    1966     4      3       5    18902   29        0        0        0   
6    2216     3      2       2    18639   29        0        0        0   
7    1597     2      1       1    22112   46        0        0        0   
8    1622     3      2       2    14321   26        0        0        0   
9    1976     3      3       1    32358   83        0        0        0   
11   2791     3      4       3    30595    9        0        0        0   
14   2090     3      3       2    24025   14        0        0        0   
16   1502     2      2       2    28958   45        0        0        0   
17   2747     3      4       2    22241    8        0        0        0   
18   2890     4      3       2    41992   47        0        0        0   
19   1825     3      3       2    30266   44        0        0        0   
20   1578     4      3       2    18829   15        1        0        0   
21   1763     3      3       2    24726   42        0        0        0   
22   2517     4      3       2    23261   36        0        0        0   
23   1872     3      2       2    24017   16        0        1        0   
24   3266     3      3       2    24881   93        0        0        0   
26   2164     4      2       2    28291   36        0        0        0   
27   2080     3      2       2    14752   42        0        0        0   
28   2208     2      2       2     8058   16        0        0        0   
30   1460     3      2       2     9999   23        0        0        0   
31   2540     3      2       2    42428   61        0        0        0   
32   2208     2      2       2     6746   16        0        0        0   
33   2560     4      2       3    36100   24        0        0        0   
34   2061     4      2       2    20138   43        0        0        0   
35   1980     3      2       2    32519   44        0        0        0   
..    ...   ...    ...     ...      ...  ...      ...      ...      ...   
487  1578     3      1       2    22485   26        0        0        0   
488  1618     3      2       2    27539   39        0        0        0   
489  1642     3      1       2    14901   42        0        0        0   
491  1604     3      2       1    23534   41        0        0        0   
492  2035     4      2       3    20131   39        0        0        0   
493  1534     3      2       1    15361   46        0        1        0   
494  1624     2      2       2    16721   46        0        0        0   
495  1700     4      2       1    15391   47        0        0        0   
496  1630     3      2       2    24963   23        1        0        0   
497  1526     3      2       1    15007   44        0        0        0   
498  1672     3      1       2    22617   52        0        0        0   
499  1588     4      2       2    21925   21        1        0        0   
500  1752     4      2       2     9126   51        0        0        0   
501  1674     4      2       2    33237   54        0        0        0   
502  1980     3      1       1    47679   74        0        0        0   
505  1263     2      2       0    12357   46        0        0        0   
507  1304     3      1       2    21305   50        0        0        0   
508  1340     3      1       0     5666   49        0        0        0   
509  1559     2      1       2    23999   49        0        0        0   
511  1198     2      2       2    20918   76        0        0        0   
512  1424     2      1       2    16414   54        0        0        0   
513  1370     4      1       1

In [118]:
y0 = df_homes.iloc[:,0]
y0[:5]

0    360000
1    340000
2    250000
3    205500
4    275500
Name: price, dtype: int64

In [119]:
#convert X_full to numpy array
X = df_homes2.to_numpy()

In [120]:
X1 = sm.add_constant(X)
X1[:5]

array([[1.0000e+00, 3.0320e+03, 4.0000e+00, 4.0000e+00, 2.0000e+00,
        2.2221e+04, 2.9000e+01, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00,
        0.0000e+00, 0.0000e+00, 1.0000e+00, 0.0000e+00],
       [1.0000e+00, 2.0580e+03, 4.0000e+00, 2.0000e+00, 2.0000e+00,
        2.2912e+04, 2.5000e+01, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00,
        0.0000e+00, 0.0000e+00, 1.0000e+00, 0.0000e+00],
       [1.0000e+00, 1.7800e+03, 4.0000e+00, 3.0000e+00, 2.0000e+00,
        2.1345e+04, 2.1000e+01, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00,
        0.0000e+00, 0.0000e+00, 1.0000e+00, 0.0000e+00],
       [1.0000e+00, 1.6380e+03, 4.0000e+00, 2.0000e+00, 2.0000e+00,
        1.7342e+04, 3.8000e+01, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00,
        0.000

In [121]:
list4.insert(0,'constant')
list4

['constant',
 'area',
 'beds',
 'baths',
 'garage',
 'lotsize',
 'age',
 'style_2',
 'style_3',
 'style_4',
 'style_5',
 'style_6',
 'style_7',
 'style_9',
 'ac_YES',
 'pool_YES',
 'quality_LOW',
 'quality_MEDIUM',
 'highway_YES']

In [123]:
X1[:5] 

array([[1.0000e+00, 3.0320e+03, 4.0000e+00, 4.0000e+00, 2.0000e+00,
        2.2221e+04, 2.9000e+01, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00,
        0.0000e+00, 0.0000e+00, 1.0000e+00, 0.0000e+00],
       [1.0000e+00, 2.0580e+03, 4.0000e+00, 2.0000e+00, 2.0000e+00,
        2.2912e+04, 2.5000e+01, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00,
        0.0000e+00, 0.0000e+00, 1.0000e+00, 0.0000e+00],
       [1.0000e+00, 1.7800e+03, 4.0000e+00, 3.0000e+00, 2.0000e+00,
        2.1345e+04, 2.1000e+01, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00,
        0.0000e+00, 0.0000e+00, 1.0000e+00, 0.0000e+00],
       [1.0000e+00, 1.6380e+03, 4.0000e+00, 2.0000e+00, 2.0000e+00,
        1.7342e+04, 3.8000e+01, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00,
        0.000

In [ ]:
#Style Base case is Style 1 which implies that a house will be of that style iff style_2,3,4,5,6,7,9 == 0.

In [124]:
m5 = sm.OLS(y0,X1).fit()

In [125]:
m5.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.852
Model:                            OLS   Adj. R-squared:                  0.845
Method:                 Least Squares   F-statistic:                     135.9
Date:                Wed, 25 Sep 2019   Prob (F-statistic):          1.08e-163
Time:                        16:24:36   Log-Likelihood:                -5451.8
No. Observations:                 445   AIC:                         1.094e+04
Df Residuals:                     426   BIC:                         1.102e+04
Df Model:                          18                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1.296e+05   2.67e+04      4.848      0.000    7.71e+04    1.82e+05
x1           117.2176      7.739     15.146      0.000     102.006     132.429
x2         -2222.9820   4148.594     -0.536      0.592   -1.04e+04    5931.279
x3          9246.5631   4285.513      2.158      0.032     823.180    1.77e+04
x4          7423.7368   4914.563      1.511      0.132   -2236.074    1.71e+04
x5             1.3283      0.238      5.584      0.000       0.861       1.796
x6         -1300.6190    199.782     -6.510      0.000   -1693.300    -907.938
x7         -2.006e+04   8638.441     -2.322      0.021    -3.7e+04   -3080.425
x8         -1.151e+04   8251.632     -1.395      0.164   -2.77e+04    4707.107
x9           2.31e+04   1.72e+04      1.346      0.179   -1.06e+04    5.68e+04
x10        -7407.6304   1.55e+04     -0.479      0.632   -3.78e+04     2.3e+04
x11         -3.06e+04   1.51e+04     -2.029      0.043   -6.02e+04    -951.905
x12        -4.664e+04   8514.836     -5.477      0.000   -6.34e+04   -2.99e+04
x13        -9.094e+04   5.26e+04     -1.728      0.085   -1.94e+05    1.25e+04
x14        -1075.0761   7501.411     -0.143      0.886   -1.58e+04    1.37e+04
x15         2.163e+04   1.06e+04      2.042      0.042     813.263    4.25e+04
x16        -1.368e+05   1.45e+04     -9.433      0.000   -1.65e+05   -1.08e+05
x17        -1.363e+05   1.08e+04    -12.640      0.000   -1.58e+05   -1.15e+05
x18         -3.18e+04   1.69e+04     -1.884      0.060    -6.5e+04    1380.295
==============================================================================
Omnibus:                       58.445   Durbin-Watson:                   1.344
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              269.978
Skew:                           0.449   Prob(JB):                     2.37e-59
Kurtosis:                       6.709   Cond. No.                     5.71e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.71e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [126]:
round(m5.params,2)

const    129624.63
x1          117.22
x2        -2222.98
x3         9246.56
x4         7423.74
x5            1.33
x6        -1300.62
x7       -20059.70
x8       -11511.87
x9        23097.27
x10       -7407.63
x11      -30595.70
x12      -46636.80
x13      -90937.91
x14       -1075.08
x15       21631.70
x16     -136786.65
x17     -136349.57
x18      -31800.51
dtype: float64

In [ ]:
list4 = ['area', 'beds', 'baths', 'garage', 'lotsize', 'age', 'style_2', 'style_3', 'style_4', 'style_5', 'style_6', 'style_7', 'style_9', 'ac_YES', 'pool_YES', 'quality_LOW', 'quality_MEDIUM', 'highway_YES']

*Fitted Equation for all columns of dataset:*

yhat = 129624.63 + 117.22 area - 2222.98 beds + 9246.56 baths + 7423 garage + 1.33 lotsize - 1300.62 age - 20059.70 style_2 - 11511.87 style_3 + 23097.27 style_4 - 7407.63 style_5 - 30595.7 style_6 - 46636.8 style_7 - 90937.91 style_9 - 1075 ac_YES - 21631.7 pool_YES -136786.65 quality_LOW - 136349.57 quality_MEDIUM - 31800.51 highway_YES

In [127]:
list_values2 = [1, 3150, 2, 3, 2, 26250, 5, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0 ,0, 0]

In [128]:
x = pd.DataFrame([list_values2], columns = list4)

In [129]:
d4 = m5.get_prediction(x)

In [130]:
#95% Prediction for the house with given feature values is  $585922.53 with prediction interval of
d4.summary_frame(alpha = .05)

mean       mean_se  mean_ci_lower  mean_ci_upper   obs_ci_lower  \
0  585922.526796  15323.494616  555803.458271  616041.595321  479849.693761   

    obs_ci_upper  
0  691995.359832

***Q8 Answer***: The 95% prediction interval for the price of this house is [ \\$479849.69,\\$691995.36 ]